### PR TITLE
Feature: Docker execution support

### PR DESCRIPTION
**Pull request by icanhazstring**
This PR ~states a Prove Of Concept for~ should solve issue #29.

The basic idea behind the changes:
- Add docker support for people (like me) that don't run their dev environment on local machine but rather on virtual ones (vagrant, docker). This has nothing to do with the ability to provide execution in docker/vagrant or anything else, this is just pure dev sugar 😉 This could also be provided in an extra PR and merged right away if this is fine

- Remove the actual hook implementation from the "hooks" it self and move them into a single point of execution (`captain-runner.php` - current working title, object to change)
- `CaptainHook\App\Hook\Template` will now contain a environment specific script for executing the `captain-runner.php`. This needs to be split up to support different kind of templates:
  - Local (for local development on "host" machine)
  - Docker (obviously docker support)
  - Vagrant (as optional later addition)
- `CaptainHook\App\Hook\Runner` was introduces which now contains the old implementation of the hooks
  - Here we have slight change to get the first argument, which is the hook that should be run
  - This gets removed, so we have the "old" behavior still intact for the app itself

Further changes that need to be done and should be discussed:
- `CaptainHook\Console\Command\Configuration` should ask what kind of environment a user is using (local, virtual)
  - If its virtual we should als for the type (docker, vagrant, what ever)
  - After we know what environment the user is using - we need to gather more information for the specific environment: For docker its the `container-name`

There are some parts in this implementation that are non functional!
Like getting the container-name it self (as stated in *what needs to be done*)

---

For easier tests, simply take the final `captain-runner.php` content which will always be the same and place them directly beside the other hooks (`.git/hooks`)

```php
<?php
$autoLoader = __DIR__ . '/../../vendor/autoload.php';

if (!file_exists($autoLoader)) {
    fwrite(STDERR,
        'Composer autoload.php could not be found' . PHP_EOL .
        'Please re-install the hook with:' . PHP_EOL .
        '$ captainhook install --composer-vendor-path=...' . PHP_EOL
    );
    exit(1);
}

$hook = $_SERVER['argv'][1];
array_splice($_SERVER['argv'], 1, 1);

require $autoLoader;
$config = realpath(__DIR__ . '/../../captainhook.json');
$app    = new CaptainHook\App\Console\Application\Hook();
$app->setHook($hook);
$app->setConfigFile($config);
$app->run();
```

### CODE REVIEW

**Comment by sebastianfeldmann** [comment]
Thanks for putting in the work for your POC.

I think I would prefer having the `hook-runner` in an actual file and not only on a template.
Putting it besides the `captainhook` binary as `captainhook-run` and registering it in the `composer.json` as binary so it will be moved to `vendor/bin` as well.

```javascript
"bin": [
    "captainhook",
    "captainhook-run"
],
```
Then add a option to the install command
`captainhook install --run-mode=local|docker|vagrant`

I'm thinking about installing `local` by default or asking the user if no option is provided.
I'm leaning more towards installing `local` by default to not bother most users with a question that is hard to phrase in a way the user understands the problem we are trying to solve.

- Do you have php installed where you execute your git commands?
- Do you run php on your local machine?
- Do you ...

I would argue none of these variants are understandable from a user standpoint. Users that are running docker or vagrant should run `captainhook install --run-mode=docker` manually.

Last thing we have to do is path the `run-mode` option down to the `writeHookFile` => `getHookSourceCode` => `Template::getCode` chain and enable the Template to return both variants, the default php version and the new docker and later vagrant versions.

### Composer Plugin Users
We should add an option to the `composer` extra section. So you can provide the `run-mode` similar to the `repository-path`. So if you are using the `plugin-composer` package you can define the `run-mode` for the whole team if you want to.

**Comment by icanhazstring** [comment]
Thanks for the reply. This is indeed a better and easier way to communicate with the user by using a run-mode.

> Last thing we have to do is pass the run-mode option down to the writeHookFile => getHookSourceCode => Template::getCode chain and enable the Template to return both variants, the default php version and the new docker and later vagrant versions.

Wouldn't it be better to have multiple template classes to separate the actual behavior?

**Comment by sebastianfeldmann** [comment]
Yes indeed that would be nicer.
And we should use an `Installer` property instad of passing the mode all the way down.

`CaptainHook\App\Runner\Installer`
```php

const RUN_MODE_LOCAL = 'local';
const RUN_MODE_DOCKER = 'docker';

...

/**
 * Run mode affects the hook implementation
 * 
 * - local   - PHP is installed locally and git hooks can be php scripts
 * - docker  - PHP is not installed locally, use docker to create a container to execute CaptainHook
 * - vagrant - [PLANNED] A Vagrant VM is available to execute the hook
 * 
 * @var string 
 */
protected $runMode;

...

/**
 * RunMode setter
 *
 * @param string $runMode
 */
public function setRunMode(string $runMode) : void
{
    $this->runMode = $runMode;
}

...
protected function getHookSourceCode(string $hook) : string
{
    $absRepoPath = realpath($this->repository->getRoot());
    $vendorPath  = getcwd() . '/vendor';
    $configPath  = realpath($this->config->getPath());
    // resolve the right template for the current run mode
    $template = $this->createTemplate();
    return $template->getCode($hook, $absRepoPath, $vendorPath, $configPath);
}

/**
 * Creates the template responsible for the current run mode
 * 
 * @return \CaptainHook\App\Hook\Template
 */
protected function createTemplate() : Template
{
    // return Template depending on $this->runMode
}
```

`CaptainHook\App\Console\Command\Install`
```php
...

->addOption(
   'run-mode',
   'r',
   InputOption::VALUE_OPTIONAL,
   'git hook run mode [local|docker]',
   Installer::RUN_MODE_LOCAL
);

...

$installer->setForce($input->getOption('force'))
          ->setHook((string) $input->getArgument('hook'))
          ->setRunMode((string) $input->getOption('run-mode');

```

**Comment by icanhazstring** [comment]
Sounds like a practical solution :+1:.
Will update the PR then.

File: captainhook-run [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+if ($hook === null) {
+    fwrite(
+        STDERR,
+        sprintf('Expected at least 3 parameters but got %d', count($_SERVER['argv']))
+    );
+    exit(1);
+}
+
+array_splice($_SERVER['argv'], 1, 1);
```
Comment (icanhazstring) [inline captainhook-run]: This is used to remove the "hook" name from `argv` so the `ArgvInput` from symfony works the same as before.

**Comment by icanhazstring** [comment]
As requested some changes in the implementation.
I also came across some missing return types on UnitTests. So I fixed them on the go.

File: tests/CaptainHook/Storage/FileTest.php [hunk 2]
```diff
@@ -92,11 +89,11 @@ public function testNoWritePermission()
...
      * Tests File::write
      */
-    public function testCantCreateDirectory()
+    public function testCantCreateDirectory(): void
     {
-        $this->expectException(\Exception::class);
+        $this->expectException(Exception::class);
 
-        $baseDir = sys_get_temp_dir() . DIRECTORY_SEPARATOR . uniqid('basedir');
+        $baseDir = sys_get_temp_dir() . DIRECTORY_SEPARATOR . uniqid('basedir', true);
```
Comment (icanhazstring) [inline tests/CaptainHook/Storage/FileTest.php]: The second parameter improves the "randomness" of `uniqid` and is recommended to be used.

**Pull request merged**